pkg/server/config.go

```diff
@@ -0,0 +1,494 @@
+	genericapiserver "k8s.io/apiserver/pkg/server"
+	"k8s.io/apiserver/pkg/util/webhook"
+	"k8s.io/client-go/dynamic"
+	coreexternalversions "k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	"k8s.io/client-go/tools/clusters"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/genericcontrolplane"
+)
+
+type Config struct {
+	options *options.CompletedOptions
+}
+
+type CompletedConfig struct {
+	EtcdServer                            *etcd.Server
+	kcpSharedInformerFactory              kcpexternalversions.SharedInformerFactory
+	kubeSharedInformerFactory             coreexternalversions.SharedInformerFactory
+	apiextensionsSharedInformerFactory    apiextensionsexternalversions.SharedInformerFactory
+	dynamicDiscoverySharedInformerFactory *informer.DynamicDiscoverySharedInformerFactory
+
+	identityConfig        *rest.Config
+	serverChain           *genericcontrolplane.ServerChain
+	server                *genericapiserver.GenericAPIServer
+	genericConfig         *genericapiserver.Config
+	preHandlerChain       handlerChainMuxes
+	profilerAddress       string
+	enabledControllers    sets.String
+	enabledAllControllers bool
+	options               *options.CompletedOptions
+
+	// TODO(sttts): get rid of these. We have wildcard informers already.
+	rootKcpSharedInformerFactory  kcpexternalversions.SharedInformerFactory
+	rootKubeSharedInformerFactory coreexternalversions.SharedInformerFactory
+}
+
+func (c *CompletedConfig) Factories() (kcpexternalversions.SharedInformerFactory, coreexternalversions.SharedInformerFactory, apiextensionsexternalversions.SharedInformerFactory, *informer.DynamicDiscoverySharedInformerFactory) {
+	return c.kcpSharedInformerFactory, c.kubeSharedInformerFactory, c.apiextensionsSharedInformerFactory, c.dynamicDiscoverySharedInformerFactory
+}
+
+func (c *CompletedConfig) IdentityConfig() *rest.Config {
+	return rest.CopyConfig(c.identityConfig)
+}
+
+func (c *CompletedConfig) ServerChain() *genericcontrolplane.ServerChain {
+	return c.serverChain
+}
+
+func (c *CompletedConfig) Server() *genericapiserver.GenericAPIServer {
+	return c.server
+}
+func (c *CompletedConfig) GenericConfig() *genericapiserver.Config {
+	return c.genericConfig
+}
+func (c *CompletedConfig) PreHandlerChain() handlerChainMuxes {
+	return c.preHandlerChain
+}
+func (c *CompletedConfig) ProfilerAddress() string {
+	return c.profilerAddress
+}
+func (c *CompletedConfig) IsControllerEnabled(name string) bool {
+	return c.enabledAllControllers || c.enabledControllers.Has(name)
+}
+func (c *CompletedConfig) VirtualEnabled() bool {
+	return c.options.Virtual.Enabled
+}
+func (c *CompletedConfig) Virtual() options.Virtual {
+	return c.options.Virtual
+}
+func (c *CompletedConfig) Controllers() options.Controllers {
+	return c.options.Controllers
+}
+func (c *CompletedConfig) Authorization() options.Authorization {
```

Comment:
all these accessors make customization hard and don't add any abstraction.

Reply:
While I do agree, these are necessary if `Config` lives outside of the same package as the thing that uses it (i.e. `Server`). `Options` live in `pkg/server/options`, so if that is one layer below `Config`, which is one layer below `Server`, wouldn't they eventually live in `pkg/server/config`? I don't mind removing them for now, but we end up with an uneven abstraction, which makes it harder for people to grasp.